sklearn/utils/__init__.py

```diff
@@ -428,24 +428,22 @@ def _get_column_indices(X, key):
         elif isinstance(key, slice):
             start, stop = key.start, key.stop
             if start is not None:
-                start = all_columns.index(start)
+                start = all_columns.get_loc(start)
             if stop is not None:
                 # pandas indexing with strings is endpoint included
-                stop = all_columns.index(stop) + 1
+                stop = all_columns.get_loc(stop) + 1
             else:
                 stop = n_columns + 1
             return list(range(n_columns)[slice(start, stop)])
         else:
             columns = list(key)
 
         try:
-            column_indices = [all_columns.index(col) for col in columns]
```

Comment:
Do we have some test, to gracefully fail with duplicate columns names, where `get_loc` wouldn't behave as expected?

Reply:
We do not have a test for this behavior. The original behavior uses `index`, which returns the first appearance of the column name.

I would consider this a bug and error.

Reply:
Updated PR with a check on the return of `get_loc`.